File: core/src/ru/myitschool/dcrawler/ai/task/SkillTask.java
package ru.myitschool.dcrawler.ai.task;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.skills.Skill;
import ru.myitschool.dcrawler.skills.Target;

/**
 * Created by Voyager on 12.02.2018.
 */
public class SkillTask extends Task {

    private Skill skill;
    private Array<Target> targets;

    public SkillTask(Entity entity, Skill skill) {
        this.skill = skill;
    }

    public SkillTask(Skill skill, Target target) {
        this(skill, new Array<Target>());
        addTarget(target);
    }

    public SkillTask(Skill skill, Array<Target> targets) {
        this.skill = skill;
        this.targets = targets;
    }

    public void addTarget(Target target){
        targets.add(target);
    }

    @Override
    protected void startTask() {
        entity.setUsedSkill(skill);
        if (targets != null){
            entity.addTargets(targets);
        }
    }

    @Override
    public void endSkill() {
        complete();
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/ai/task/SkillTask.kt
package com.wolg_vlad.dcrawler.entities.ai.task

import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.Skill
import com.wolg_vlad.dcrawler.entities.skills.Target

/**
 * Created by Voyager on 12.02.2018.
 */
class SkillTask : Task {
    private var skill: Skill
    private var targets = mutableListOf<Target>()

    constructor(entity: Entity?, skill: Skill) {
        this.skill = skill
    }

    constructor(skill: Skill, target: Target) : this(skill, mutableListOf()) {
        addTarget(target)
    }

    constructor(skill: Skill, targets: List<Target>) {
        this.skill = skill
        this.targets = targets.toMutableList()
    }

    fun addTarget(target: Target) {
        targets!!.add(target)
    }

    override fun startTask() {
        entity!!.setUsedSkill(skill)
        entity!!.addTargets(targets)
    }

    override fun endSkill() {
        complete()
    }
}